src/radio_client.c

```diff
@@ -0,0 +1,1126 @@
+    while (g_hash_table_iter_next(&it, NULL, &value)) {
+        RadioRequest* req = value;
+
+        if (req->group != group) {
+            /*
+             * There's a pending request not associated with any group
+             * or associated with a different group. The specified group
+             * can't become the owner just yet.
+             */
+            return FALSE;
+        }
+    }
+
+    /*
+     * There are no pending requests, or all pending requests are
+     * associated with the specified group. This group can be the owner.
+     */
+    return TRUE;
+}
+
+static
+void
+radio_client_queue_request(
+    RadioClientObject* self,
+    RadioRequest* req)
+{
+    req->state = RADIO_REQUEST_STATE_QUEUED;
+    if (self->queue_last) {
+        self->queue_last->queue_next = req;
+    } else {
+        self->queue_first = req;
+    }
+    self->queue_last = req;
+}
+
+static
+guint
+radio_client_submit_queued_requests(
+    RadioClientObject* self)
+{
+    guint submitted = 0;
+
+    if (!self->block_req && self->pub.instance->connected) {
+        RadioRequest* prev = NULL;
+        RadioRequest* req = self->queue_first;
+
+        while (req) {
+            if (self->owner == req->group ||
+               (!self->owner && !self->owner_queue)) {
+                /* Remove it from the queue */
+                radio_client_dequeue_request(self, req, prev);
+
+                /* Initiate the transaction and update the request state */
+                if (radio_client_submit_transaction(self, req)) {
+                    submitted++;
+                    if (req->blocking) {
+                        self->block_req = radio_request_ref(req);
+                        break;
+                    }
+                } else {
+                    radio_client_request_failed(self, req);
+                }
+            } else {
+                prev = req;
+            }
+            req = req->queue_next;
+        }
+    }
+    return submitted;
+}
+
+static
+void
+radio_client_handle_death(
+    RadioInstance* instance,
+    gpointer user_data)
+{
+    RadioClientObject* self = THIS(user_data);
+    RadioRequest* dead = NULL;
+    RadioRequest* req;
+    GHashTableIter it;
+    gpointer value;
+
+    g_object_ref(self);
+
+    /* Steal all active requests from the table */
+    g_hash_table_remove_all(self->pending);
+    g_hash_table_iter_init(&it, self->active);
+    while (g_hash_table_iter_next(&it, NULL, &value)) {
+        req = value;
+        GDEBUG("Dropping request %u (%08x/%08x) due to radio death",
+            req->code, req->serial, req->serial2);
+
+        /* Don't unref them just yet */
+        g_hash_table_iter_steal(&it);
+
+        /* We can safely reuse queue_next link */
+        req->queue_next = dead;
+        dead = req;
+    }
+
+    while (dead) {
+        /* Fail the dead request and drop the stolen ref */
+        req = dead;
+        dead = req->queue_next;
+        radio_client_request_failed(self, req);
```

Comment:
Would it make sense to set `queue_first` to NULL at some point, so that deactivate_request doesn't unnecessarily try to find the req from the queue.

Reply:
Good point

Reply:
Actually, I'm only failing pending requests here, and the PENDING state stops `radio_client_deactivate_request()` from trying to find the request in the queue, so that's not a problem. But what's missing then is a flush of the queue. Will fix and test that.

Reply:
Hmm, this code is killing all active requests, including the queued ones. Then reusing `queue_next` link is plain wrong. Yeah, good catch.

Reply:
I think it's fixed now. Thanks!